### ISSUE TITLE
Issue in parallel for periodic constraints

### ISSUE THREAD

**Issue by cmaurini** [jorgensd/dolfinx_mpc#15]
I get what I think to be a bug in multipoint constraint when running in parallel. 

The example you find below give correct results on one processor bug wrong results with more then one processors. 

You find attached an image when running with `mpiexec -n 2` and the following error `mpiexec -n 3`

```
(dolfinx) inclusion_2d_elastic|main⚡ ⇒ mpiexec -n 3 python elasticity_test.py
Traceback (most recent call last):
  File "/Users/maurini/Documents/codes/FFT-FEM/fem/inclusion_2d_elastic/elasticity_test.py", line 72, in <module>
Traceback (most recent call last):
  File "/Users/maurini/Documents/codes/FFT-FEM/fem/inclusion_2d_elastic/elasticity_test.py", line 72, in <module>
    mpc.create_periodic_constraint_geometrical(V.sub(1), bottom_boundary, mapping_1, [], -1)
  File "/Users/maurini/dev/dolfinx/lib/python3.9/site-packages/dolfinx_mpc/multipointconstraint.py", line 158, in create_periodic_constraint_geometrical
    mpc.create_periodic_constraint_geometrical(V.sub(1), bottom_boundary, mapping_1, [], -1)
  File "/Users/maurini/dev/dolfinx/lib/python3.9/site-packages/dolfinx_mpc/multipointconstraint.py", line 158, in create_periodic_constraint_geometrical
    mpc_data = dolfinx_mpc.cpp.mpc.create_periodic_constraint_geometrical(
ValueError: vector
    mpc_data = dolfinx_mpc.cpp.mpc.create_periodic_constraint_geometrical(
ValueError: vector
```

You can find a script and a mesh to reproduce here [Archive.zip](https://github.com/jorgensd/dolfinx_mpc/files/8191400/Archive.zip). I do not get the error with a structured `unit_square_mesh`. The script is reproduced below. It imposed periodic bcs on a square.

```
import numpy as np
from pathlib import Path
from mpi4py import MPI
from dolfinx.common import Timer, list_timings, TimingType
from dolfinx.io import XDMFFile
from dolfinx.fem import (
    FunctionSpace,
    Function,
    Constant,
)
import ufl
import dolfinx.geometry
import dolfinx_mpc
from dolfinx_mpc.utils import rigid_motions_nullspace

outdir = "output"
comm = MPI.COMM_WORLD
if comm.rank == 0:
    Path(outdir).mkdir(parents=True, exist_ok=True)

with XDMFFile(comm, 'meshes/inclusion-1.20-1.20-0.6-0.6-0.030.xdmf', "r") as file:
    mesh = file.read_mesh()
    mesh.topology.create_connectivity(1, 2)
    mesh.topology.create_connectivity(0, 1)

Lx, Ly = 1.2, 1.2
element = ufl.VectorElement("Lagrange", mesh.ufl_cell(), 1)
V = FunctionSpace(mesh, element)
u = dolfinx.fem.Function(V)
u_test, u_trial = ufl.TrialFunction(V), ufl.TestFunction(V)
f = Constant(mesh, (0.0, 0.0))
eps_0 = Constant(mesh, -0.1)
Id = ufl.Identity(2)

def eps(u):
    return ufl.sym(ufl.grad(u))

def sigma(eps):
    lmbda = 0.0
    mu = 1.0 
    return 2.0 * mu * ufl.sym(eps) + lmbda * ufl.tr(eps) * Id


a_form = ufl.inner(sigma(eps(u_trial)), eps(u_test)) * ufl.dx
L_form = ufl.dot(f, u_test) * ufl.dx + ufl.inner(sigma(eps_0 * Id), eps(u_test)) * ufl.dx

def left_boundary(x):
    tol = 1e-5
    return np.logical_and(np.isclose(x[0], -Lx), np.logical_and(x[1] > - Ly - tol, x[1] < Ly - tol))

def bottom_boundary(x):
    tol = 1e-5
    return np.logical_and(np.isclose(x[1], -Ly), np.logical_and(x[0] > - Lx + tol, x[0] < Lx - tol))

def mapping_0(x):
    vals = np.zeros(x.shape)
    vals[0] = x[0] + 2 * Lx 
    vals[1] = x[1]
    vals[2] = x[2]
    return vals

def mapping_1(x):
    vals = np.zeros(x.shape)
    vals[0] = x[0]
    vals[1] = x[1] + 2 * Ly
    vals[2] = x[2]
    return vals

with Timer("~MPC: Initialize MPC"):
    mpc = dolfinx_mpc.MultiPointConstraint(V)
    mpc.create_periodic_constraint_geometrical(V.sub(0), left_boundary, mapping_0, [], -1)
    mpc.create_periodic_constraint_geometrical(V.sub(1), bottom_boundary, mapping_1, [], -1)
    mpc.finalize()

# Solve Linear problem
problem = dolfinx_mpc.LinearProblem(a_form, L_form, mpc, bcs=[], u=u)
null_space = rigid_motions_nullspace(mpc.function_space)
problem.A.setNullSpace(null_space)
problem.A.setNearNullSpace(null_space)
solver = problem.solver
with Timer("~MPC: Solve problem"):
    problem.solve()

with Timer("~Interpolate stress"):
    T = dolfinx.fem.TensorFunctionSpace(mesh, ("DG", 0), shape=(2,2))
    sigma_expr = dolfinx.fem.Expression(sigma(eps(u) - eps_0 * Id), T.element.interpolation_points)
    sigma_sol = Function(T,name="stress")
    sigma_sol.interpolate(sigma_expr)

u.name = "u"
with XDMFFile(MPI.COMM_WORLD, f"{outdir}/u_per.xdmf", "w") as file:
    file.write_mesh(mesh)
    file.write_function(u)
    file.write_function(sigma_sol)

list_timings(MPI.COMM_WORLD, [TimingType.wall])
```

![fig](https://user-images.githubusercontent.com/8680920/156898758-fb50f741-862b-457e-8dca-3c6e460649c0.jpg)
[Archive.zip](https://github.com/jorgensd/dolfinx_mpc/files/8191400/Archive.zip)

**Comment by jorgensd** [comment]
So first of all, could you try removing the input `u=u` to your `LinearProblem`. As I mentioned in the PR, it is important that it is initialized with the `mpc.function_space` and not the original function space (which is likely the cause of your error).

I am not getting any errors running this on my system without u=u.

**Comment by cmaurini** [comment]
Thanks, I will check and get back. This is probably the cause of the error. It was far from clear to me that the initial function space is different from mpc.function_space. I am using the dolfinx_mpc.LinearProblem inside a fixed point algorithm, that's why I need to provide my own u.

**Comment by jorgensd** [comment]
> Thanks, I will check and get back. This is probably the cause of the error. It was far from clear to me that the initial function space is different from `mpc.function_space`. I am using the `dolfinx_mpc.LinearProblem` inside a fixed point algorithm, that's why I need to provide my own `u`.

Ill think about how to make that clear through demos and documentation.

**Comment by cmaurini** [comment]
uhm ... what you suggest solve the thing. However, I need the solution in the same function space where I define it. I think this is generally desirable. Is there a way to get it?

**Comment by jorgensd** [comment]
> uhm ... what you suggest solve the thing. However, I need the solution in the same function space where I define the problem. I think this is generally desirable. Is there a way to get it?

You would need to give me some more context here. Why cant you extract u after setting up the linearproblem (before solving)?
I.e.
```
problem = dolfinx_mpc.LinearProblem(a_form, L_form, mpc, bcs=[])
u=problem.u
```

The MPC function space carries extra information about degrees of freedom to enable a local assembly process (by adding extra ghost dofs for those degrees of freedom involved in a constraint, but that is not on the current process after initially partitioning the mesh).

**Comment by cmaurini** [comment]
I have a fixed point algorithm for a nonlinear problem involving two variables `(u,v)`. I want to define `u` at the initialisation of the fixed point and update `u` at each iteration by doing `problem.solve()`, or similar, with `problem = dolfinx_mpc.LinearProblem(a_form, L_form, mpc)`. 

I tried to define at the beginning `u` as `u = Function(V_u_mpc)` with something like
```
V_u = FunctionSpace(mesh, element_u)
mpc = dolfinx_mpc.MultiPointConstraint(V_u)
mpc.create_periodic_constraint_geometrical(V_u.sub(0), right_boundary, mapping_0, [])
mpc.finalize()
V_u_mpc = mpc.function_space
u = Function(V_u_mpc)
```
and then use this `u` in my fixed point, but this does not work (gives a wrong solution suggesting a ghost-update-like issue, as above).

**Comment by jorgensd** [comment]
What about:
```
V_u = FunctionSpace(mesh, element_u)
mpc = dolfinx_mpc.MultiPointConstraint(V_u)
mpc.create_periodic_constraint_geometrical(V_u.sub(0), right_boundary, mapping_0, [])
mpc.finalize()
V_u_mpc = mpc.function_space
u = Function(V_u_mpc)
problem = dolfinx_mpc.LinearProblem(a_form, L_form, mpc)
problem.u = u
```
until ive added an interface to LinearProblem that takes in u and checks that it from the extended space?

**Comment by cmaurini** [comment]
I have done this, but still I get wrong results in parallel, see below (results are good on 1 proc)
![u](https://user-images.githubusercontent.com/8680920/157129425-548119e5-6714-4e13-9a6b-aff52304b03e.jpg)

**Comment by jorgensd** [comment]
Is this the result of the code you initially posted in the issue, or from the fixed point iteration?

**Comment by cmaurini** [comment]
Fixed point 

Corrado Maurini, Handheld device

**Comment by jorgensd** [comment]
> Fixed point Corrado Maurini, Handheld device
> […](#)

I would need to see the full fixed point iteration code to give guidance as to what has gone wrong.

**Comment by cmaurini** [comment]
Finally after carefully checking I solved my issue using 
```
V_u_mpc = mpc.function_space
u = Function(V_u_mpc)
```

**Comment by jorgensd** [comment]
I’ll keep this open as a reminder to review linearproblem

**Comment by jorgensd** [comment]
Resolved in: fa124eb77ccefe45451f1202b64a551ece97b662
@cmaurini  do you mind testing the interface?

**Comment by cmaurini** [comment]
Great, it works nicely. Thanks!
